### PR TITLE
Add `Dataset.from_generator`

### PR DESCRIPTION
**Pull request by mariosasko**
Add `Dataset.from_generator` to the API to allow creating datasets from data larger than RAM. The implementation relies on a packaged module not exposed in `load_dataset` to tie this method with `datasets`' caching mechanism.

Closes https://github.com/huggingface/datasets/issues/4417

### CODE REVIEW

**Comment by lhoestq** [comment]
I restarted the builder PR job just in case

**Comment by mariosasko** [comment]
CI is now green. https://github.com/huggingface/doc-builder/pull/296 explains why it failed.

File: docs/source/loading.mdx [hunk 1]
```diff
@@ -220,6 +220,22 @@ Load a list of Python dictionaries with [`~Dataset.from_list`]:
...
+### Python generator
+
+Create a dataset from a Python generator with [`~Dataset.from_generator`]:
+
+```py
+>>> from datasets import Dataset
+>>> def my_gen():
+...     yield {"a": 1}
+...     yield {"a": 2}
+...     yield {"a": 3}
```
Comment (lhoestq) [inline docs/source/loading.mdx:232]: (nit) you can use a for loop instead, most users will probably use a for loop IMO

File: src/datasets/packaged_modules/generator/generator.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+from dataclasses import dataclass
+from typing import Callable, Optional
+
+import datasets
+
+
+@dataclass
+class GeneratorConfig(datasets.BuilderConfig):
+    generator: Optional[Callable] = None
```
Comment (lhoestq) [inline src/datasets/packaged_modules/generator/generator.py:9]: Maybe you don't need to set None by default ?

File: src/datasets/arrow_dataset.py [hunk 3]
```diff
@@ -939,6 +939,45 @@ def from_csv(
...
+        generator: Callable,
+        features: Optional[Features] = None,
+        cache_dir: str = None,
+        keep_in_memory: bool = False,
+        gen_kwargs: Optional[dict] = None,
+    ):
+        """Create a Dataset from a generator.
+
+        Args:
+            generator (:obj:`Callable`): A callable object that returns an object that supports the `iter` protocol.
```
Comment (lhoestq) [inline src/datasets/arrow_dataset.py:953]: (nit) suggestion for the docstring
```suggestion
            generator (:obj:`Callable`): A generator function that `yields` examples.
```

**Review by lhoestq — APPROVED** [review]
Awesome thank you !

Maybe let's open an issue to also add IterableDataset.from_generator ?

File: src/datasets/packaged_modules/generator/generator.py (hunk 2 above)
Comment (mariosasko) [reply src/datasets/packaged_modules/generator/generator.py:9]: I would expect this to work, but it doesn't due to the class inheritance (see https://bugs.python.org/issue39300 for the explanation)

**Pull request merged**